src/mappings/poktroll/reports.ts

```diff
@@ -0,0 +1,314 @@
+
+    if (!relaysByServiceMap[relay.serviceId]) {
+      relaysByServiceMap[relay.serviceId] = {
+        tokens: BigInt(0),
+        computedUnits: BigInt(0),
+        amount: 0,
+        relays: BigInt(0),
+      }
+    }
+
+    relaysByServiceMap[relay.serviceId].amount += 1
+    relaysByServiceMap[relay.serviceId].relays += relay.numRelays || BigInt(0)
+    relaysByServiceMap[relay.serviceId].tokens += relay.claimedAmount || BigInt(0)
+    relaysByServiceMap[relay.serviceId].computedUnits += relay.numClaimedComputedUnits || BigInt(0)
+  }
+
+  const relaysByService = Object.entries(relaysByServiceMap).map(([service, {amount, computedUnits, relays, tokens}]) => ({
+    service,
+    tokens,
+    amount,
+    computedUnits,
+    relays
+  }))
+
+  return {
+    relays: relaysAmount,
+    computedUnits,
+    claimedAmount,
+    relaysByService
+  }
+}
+
+async function getTransactionsData(block: CosmosBlock){
+  const transactions = await Transaction.getByBlockId(block.block.id, {})
+
+  let validTxs = 0, invalidTxs = 0;
+
+  for (const tx of transactions) {
+    if (tx.code === 0) {
+      validTxs += 1
+    } else {
+      invalidTxs += 1
+    }
+  }
+
+  return {
+    validTxs,
+    invalidTxs
+  }
+}
+
+async function getStakedSuppliersData() {
+  const stakedSuppliers = await Supplier.getByFields([["stakeStatus", "=", StakeStatus.Staked]], {})
+  const stakedSuppliersByServiceMap: Record<string,{
+    tokens: bigint,
+    amount: number,
+  }> = {}
+
+  let stakedTokensBySupplier = BigInt(0)
+
+  for (const supplier of stakedSuppliers) {
+    stakedTokensBySupplier += supplier.stakeAmount
+
+    const services = await SupplierServiceConfig.getBySupplierId(supplier.id, {})
+
+    for (const {serviceId} of services) {
+      if (!stakedSuppliersByServiceMap[serviceId]) {
+        stakedSuppliersByServiceMap[serviceId] = {
+          tokens: BigInt(0),
+          amount: 0,
+        }
+      } else {
+        stakedSuppliersByServiceMap[serviceId].tokens += supplier.stakeAmount
+        stakedSuppliersByServiceMap[serviceId].amount += 1
+      }
+    }
+  }
+
+  const stakedSuppliersByService = Object.entries(stakedSuppliersByServiceMap).map(([service, {amount, tokens}]) => ({
+    service,
+    tokens,
+    amount,
+  }))
+
+  return {
+    stakedSuppliers: stakedSuppliers.length,
+    stakedTokensBySupplier,
+    stakedSuppliersByService
+  }
+}
+
+async function getUnstakingSuppliersData() {
+  const unstakingSuppliers = await Supplier.getByFields([["stakeStatus", "=", StakeStatus.Unstaking]], {})
+  const unstakingTokensBySupplier = unstakingSuppliers.reduce((acc, supplier) => acc + BigInt(supplier.stakeAmount), BigInt(0))
+
+  return {
+    unstakingSuppliers: unstakingSuppliers.length,
+    unstakingTokensBySupplier,
+  }
+}
+
+async function getTook(block: CosmosBlock) {
+  if (block.header.height === 1) {
+    return 0
+  }
+
+  const previousHeight = BigInt(block.header.height - 1)
```

Comment:
what if block is 1? this will look for block 0 and then next line will fail? maybe I miss something in the path to here.

Reply:
It will return 0. You can see an if above of this doing it.